### PR TITLE
Improve key information for Filter

### LINKED ISSUE #9397: Missing unique key information after projection
### What version of Materialize are you using?

From the shell:

```
$ materialized -v

```

Or from the SQL console:

```
materialize=> select mz_version();

```

### How did you install Materialize?

* [ ] Docker image
* [ ] Linux release tarball
* [ ] APT package
* [ ] macOS release tarball
* [ ] Homebrew tap
* [X] Built from source
* [ ] Materialize Cloud

### What was the issue?

```
materialize=> create materialized view v1 as select f1, f2 from t1 group by f1, f2;
CREATE VIEW
materialize=> explain typed select * from v1;
          Optimized Plan           
-----------------------------------
 %0 =                             +
 | Get materialize.public.v1 (u43)+
 | | types = (integer?, integer?) +
 | | keys = ((#0, #1))            +
 
(1 row)

materialize=> explain typed select f1 from v1 where f1 = f2;
          Optimized Plan           
-----------------------------------
 %0 =                             +
 | Get materialize.public.v1 (u43)+
 | | types = (integer?, integer?) +
 | | keys = ((#0, #1))            +
 | Filter (#0 = #1)               +
 | | types = (integer, integer)   +
 | | keys = ((#0, #1))            +
 | Project (#0)                   +
 | | types = (integer)            +
 | | keys = ()                    +
 
(1 row)

materialize=>
```

The tuples returned by the last query in the snippet above are guaranteed to be unique, but we fail to report that fact.

Spotted while working on #9354. See https://github.com/MaterializeInc/materialize/pull/9354#issuecomment-984483560

In summary, `Filter` should report `[0]` and `[1]` as its keys, instead of `[0, 1]`.

### PR DESCRIPTION
**Pull request by asenac**
### Motivation

This PR fixes a recognized bug.

Fixes #9397.

Use equality predicates to refine and extend the unique keys reported by `Filter`s.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

File: src/expr/src/relation/mod.rs [hunk 1]
```diff
@@ -439,6 +439,81 @@ impl MirRelationExpr {
...
+                            let mut new_keys = HashSet::new();
+                            for key in prev_keys.into_iter() {
+                                let contains_c1 = key.contains(&c1);
+                                let contains_c2 = key.contains(&c2);
+                                if contains_c1 && contains_c2 {
+                                    new_keys.insert(
+                                        key.iter()
+                                            .filter(|c| **c != c1)
+                                            .cloned()
+                                            .sorted()
```
Comment (wangandi) [inline src/expr/src/relation/mod.rs:482]: The `sorted` here is unnecessary, no? The filter would not cause the order of the retained items to change.

**Review by wangandi — APPROVED** [review]

File: src/expr/src/relation/mod.rs (hunk 1 above)
Comment (asenac) [reply src/expr/src/relation/mod.rs:482]: Good point.

**Comment by philip-stoev** [comment]
I would like to push a few more tests for this can you please wait for me until Monday? Thanks!

**Comment by philip-stoev** [comment]
I have trouble triggering this optimization ... Maybe I am misunderstanding the intent of it. A simple query that shows it in action would be much appreciated! In particular, I tried this, and the Distinct was unaffected:

```
materialize=> explain select distinct f1 from t1 where f1 = 1;
          Optimized Plan          
----------------------------------
 %0 =                            +
 | Get materialize.public.t1 (u9)+
 | Filter (#0 = 1)               +
 | Project ()                    +
 | Distinct group=()             +
 | Map 1                         +
```

**Comment by wangandi** [comment]
Based on the linked issue, I believe these should show the optimization in action:
```
create materialized view v1 as select f1, f2 from t1 group by f1, f2;
explain typed select distinct f1 from v1 where f1 = f2;
```

```
create materialized view v1 as select f1, f2 from t1 group by f2;
explain typed select distinct f1 from v1 where f1 = f2;
```

**Comment by asenac** [comment]
Yes, Andi's first example is correct (the second can't project f1 since it doesn't appear in the grouping key).

This is main:

```
materialize=> explain typed select distinct f1 from v1 where f1 = f2;
          Optimized Plan           
-----------------------------------
 %0 =                             +
 | Get materialize.public.v1 (u51)+
 | | types = (integer?, integer?) +
 | | keys = ((#0, #1))            +
 | Filter (#0 = #1)               +
 | | types = (integer, integer)   +
 | | keys = ((#0, #1))            +
 | Project (#0)                   +
 | | types = (integer)            +
 | | keys = ()                    +
 | Distinct group=(#0)            +
 | | types = (integer)            +
 | | keys = ((#0))                +
 
(1 row)

materialize=>
```

With this optimization:

```
materialize=> explain typed select distinct f1 from v1 where f1 = f2;
           Optimized Plan           
------------------------------------
 %0 =                              +
 | Get materialize.public.v1 (u102)+
 | | types = (integer?, integer?)  +
 | | keys = ((#0, #1))             +
 | Filter (#0 = #1)                +
 | | types = (integer, integer)    +
 | | keys = ((#0), (#1))           +
 | Project (#0)                    +
 | | types = (integer)             +
 | | keys = ((#0))                 +
 
(1 row)

materialize=> 
```

ie. the distinct operator in the query can be removed since the uniqueness is guaranteed by the equality predicate between two columns belonging to a unique key.

**Comment by philip-stoev** [comment]
Item No #1  - @wangandi thanks. So the example above is different because it would require that key information is generated, as opposed to propagated:

```
materialize=> explain typed plan for select distinct f1 from t1 where f1 = 1;
                  Optimized Plan                   
---------------------------------------------------
 %0 =                                             +
 | Get materialize.public.t1 (u25)                +
 | | types = (double precision?, double precision)+
 | | keys = ()                                    +
 | Filter (#0 = 1)                                +
 | | types = (double precision, double precision) +
 | | keys = () <- this should not have been empty.
```
but it does not. @asenac would that be within scope of this PR?

**Comment by asenac** [comment]
@philip-stoev  Why should that not be empty? `t1` has no unique keys, so we can neither propagate them nor refine them through equalities.

**Comment by philip-stoev** [comment]
@asenac Yes now that I thought of it again, f1 = 1 does not give you anything in terms of uniqueness. So please disregard that entire thing. Seems to happen quite frequently for me those days to miss obvious stuff ...

Anyway, give me a day or two to compose and push a few more scenarios with full EXPLAINs for my own satisfaction.

**Review by philip-stoev — APPROVED** [review]
I pushed an additional SLT file in your branch -- can you please squash it along with the rest of the fix?

No issues found, the optimization kicks in as expected.

**Pull request merged**